feat: add annotation schema tab to annotation panel and editable properties for annotations

seankmartin opened the pull request:
Readonly datasources get a readonly view of the schema. Editable datasources (e.g. local annotation layers) get an editable version of the schema to add properties to. Editable datasources can also modify properties of annotations from the selection panel.

<img width="477" height="490" alt="image" src="https://github.com/user-attachments/assets/e0ef5120-e95c-44a2-af91-a41563ce61b7" />

<img width="483" height="405" alt="image" src="https://github.com/user-attachments/assets/e7885f66-73d6-4a11-93cc-b42bbde930f8" />

seankmartin committed:
feat: add editable annotation schema and properties

Co-authored-by: Aigul Zhyldyzbekova <aiguljhyldyzbekkyzy@gmail.com>
Co-authored-by: Chris Jordan <jordanchriss@gmail.com>

seankmartin commented:
Thanks for the suggestions! I opened the first as an issue with some ideas for now

seankmartin committed:
fix: correct type cell height

seankmartin committed:
refactor: don't add remove class list in css

seankmartin committed:
fix: don't allow multiple dropdown menus to be open

seankmartin committed:
refactor: don't rebuild type add dropdown each time

seankmartin committed:
style: lint and format

seankmartin committed:
Merge remote-tracking branch 'upstream/master' into feature/annotation-schema

seankmartin committed:
feat: make bool a real annotation type

seankmartin committed:
style: lint and format

seankmartin committed:
Merge pull request #112 from MetaCell/feat/bool-annotation-type

feat: make bool a real annotation type

seankmartin committed:
fix: watch for scroll on color

seankmartin committed:
feat: ignore annotation view updates from property changes

seankmartin committed:
feat: add signal so selection state updates

seankmartin committed:
feat: scaffolding an enum default select

seankmartin committed:
fix: ensure to always go back to true on allow update

seankmartin committed:
style: clarify comment and styling

seankmartin committed:
Merge pull request #113 from MetaCell/fix/editable-annotation-updates

fix: watch for scroll on color

seankmartin committed:
Merge branch 'feature/annotation-schema' into fix/enum-default-select

seankmartin committed:
feat: hook up selector properly for default

seankmartin committed:
fix: don't update default on enums in UI

seankmartin committed:
fix: allow schema to move in default

seankmartin committed:
feat: explain default updates

seankmartin committed:
fix: small css fixes

seankmartin committed:
Merge pull request #114 from MetaCell/fix/enum-default-select

feat: create an enum default select

seankmartin committed:
fix: correct enum name input overwrite value input

seankmartin committed:
fix: show non schema enum values

seankmartin committed:
fix: correct enum default value interaction

seankmartin committed:
fix: extend float step

seankmartin committed:
fix: update enum default select options

seankmartin committed:
fix: correct updating of enum values

seankmartin committed:
fix: small css improvements

seankmartin committed:
fix(test): correct failing test

seankmartin committed:
fix: show default on readonly enum

seankmartin committed:
chore: remove handled TODO

seankmartin committed:
Merge remote-tracking branch 'origin/master' into feature/annotation-schema

seankmartin committed:
fix: update polyline for editable props

seankmartin committed:
fix: allow default props to be readonly

seankmartin committed:
fix: annotation property editing accounts for possible defaults in list

fcollman commented:
@seankmartin I think you need to rebase to get the mac tests to pass.. @chrisj has got the CLA agreement fixed.

seankmartin committed:
Merge branch 'master' into feature/annotation-schema

seankmartin committed:
Merge branch 'master' into feature/annotation-schema

chrisj commented:
The schema tab feels great and feels like a naturall fit!

<img width="223" height="174" alt="Screenshot 2026-04-07 at 1 19 09 PM" src="https://github.com/user-attachments/assets/80a89e81-6d63-47f0-852e-ac95a4433d6a" />

It seems the change type dropdown is incorrectly showing the numeric property icon instead of the enum icon when changing an enum.

The dropdown is also missing two things that seem to be working for the add property dropdown. It doesn't flip direction if it would otherwise overflow off the bottom of the screen and it doesn't have the same hover highlight.

chrisj commented:
The paste from clipboard feature is very convenient though I was a little unsure how to use it at first. Now I know, I wouldn't suggest any changes. I would add an upload schema action to match the corresponding download.

seankmartin committed:
fix: run enum labels through to type change for correct icon

seankmartin committed:
fix: move custom event to hover css for consistency
and more natural

seankmartin committed:
refactor: rename confusing dropdowns

seankmartin committed:
fix: use common placement on type change and add dropdown

seankmartin committed:
feat: add annotation schema uploader

seankmartin committed:
merge: merge master, fix conflict on elipses overflow

seankmartin commented:
thanks @chrisj! added/fixed those

chrisj commented on src/annotation/index.ts:
```diff
@@ -1479,14 +1572,112 @@ export class LocalAnnotationSource extends AnnotationSource {
+
+    const convertValue = (value: any) => {
+      if (oldType === "rgb" && newType === "rgba") {
+        const rgba = new Uint8Array(4);
```
The value here seems to be a packedRGB so you need to unpack and repack it. Converting a property from RGB to RGBA destroys the values.

seankmartin replied:
Thank you for spotting that Chris, I wrote an extender instead of unpacking and repacking but same thing really so hopefully that fixes it. Cheers!

chrisj replied:
even better!

chrisj commented:
AI analysis is concerned about some situations with multiple annotation sources per layer, particularly if some are editable and some are read only. I'm still figuring out if any of them are real/reproducible.

seankmartin committed:
fix: pack alpha onto color value in rgb -> rgba conversion

seankmartin committed:
refactor: move color extension to color.ts

seankmartin commented:
> AI analysis is concerned about some situations with multiple annotation sources per layer, particularly if some are editable and some are read only. I'm still figuring out if any of them are real/reproducible.

Thanks Chris, I think that's likely an issue, as admittedly it wasn't really a situation thought through. I can, for example, add a local annotation source to an annotation layer and then because one of the sources is editable even though the source doesn't match in properties - the annotation schema now thinks the whole schema is editable. Of course the edits don't make it through in most cases because the most common type (precomputed) doesn't support writing. But it's definitely an issue if more sources can be read/write, a source declares itself to be in read only mode, and then you add a local annotation source to the layer and now you can write schema changes to the initial source.

To have two annotation sources in a layer their properties need to be an exact match. However, for local annotations the annotation properties in the state are stored at the layer level not the source level. I think the safest bet for now would be to change the logic on the editable check to check if all annotation sources in the layer are editable. Right now it checks if any source is editable.

Very happy to discuss alternatives, especially if that blocks some kind of workflow I'm missing

seankmartin committed:
fix: protect against a single editable annotation source making readonly sources
editable

protects against sn issue if more sources can be read/write, a source declares itself to
be in read only mode, and then you add a local annotation source to the layer and now
you can write schema changes to the initial source.

chrisj commented:
@seankmartin thanks for these updates! A few remaining issues.

* The upload schema button you added at my request is showing up for readonly sources.
* Annotation selection details panel does not update when the schema changes.

chrisj commented on src/ui/annotations.ts:
```diff
@@ -1986,15 +2007,38 @@ export function UserLayerWithAnnotationsMixin<
+      const throttledUpdate = context.registerCancellable(
+        throttle(() => {
+          if (!newAnnotation) return;
+          this.allowDependentAnnotationViewUpdate.value = false;
```
I would add a little comment here because it was hard for me to follow but I understand it as:

When editing an annotation from the ui controls within the selection details, we don't need to update the selection details since the user input already performed the necessary update. This is important for at least the color picker control because otherwise it causes the color picker to immediately close after changing the value.

seankmartin committed:
fix: hide upload schema on readonly

seankmartin committed:
docs: add comment explaining throttling + view update blocking

seankmartin committed:
fix: add layer update to property update
This allows the selection details to update

seankmartin commented:
> * Annotation selection details panel does not update when the schema changes.

Thanks for spotting these issues! Didn't realise I'd forgotten to add a call to the signal that the layers changed on property update (was present for the property add/remove)
